SeaBattle/Ships/BaseShip.cs

```diff
@@ -0,0 +1,159 @@
+        // 4. + checking if speed is enough to get to new position
+        // 5. + checking if new position is available
+        // 6. + updating current position with new position
+        public void MoveTo(List<Point> NewPosition, Sea Sea1)
+        {
+            float midShipX = 0;
+            float midShipY = 0;
+            foreach (var d in Decks)
+            {
+                midShipX += d.Location.X;
+                midShipY += d.Location.Y;
+            }
+            midShipX /= Decks.Count;
+            midShipY /= Decks.Count;
+
+            float midNewX = 0;
+            float midNewY = 0;
+            foreach (var p in NewPosition)
+            {
+                midNewX += p.X;
+                midNewY += p.Y;
+            }
+            midNewX /= NewPosition.Count;
+            midNewY /= NewPosition.Count;
+
+            float TravelX = (float)Math.Pow(midShipX - midNewX, 2);
+            float TravelY = (float)Math.Pow(midShipY - midNewY, 2);
+            float TravelDistance;
+            TravelDistance = (float)Math.Sqrt(TravelX + TravelY);
+            if (Speed < TravelDistance)
+            {
+                throw new Exception("Ship can't move");
+            }
+            if (Sea1.IsPlaceAvailable(this) && this.ShipCoordsLine(NewPosition) && this.ShipCoordsUnion(NewPosition) && this.ShipCoordsUnique(NewPosition))
+            {
+                for (int i = 0; i < NewPosition.Count; i++)
+                {
+                    Decks[i].Location.X = NewPosition[i].X;
+                    Decks[i].Location.Y = NewPosition[i].Y;
+                }
+            }
+            else
+                throw new Exception("Ship can't move");
+        }
+        public bool HasCoordinatesIntersect(List<Deck> DecksList)
+        {
+            return Decks.Any(e => e.HasCoordinatesIntersect(DecksList));
+        }
+        public bool ShipCoordsLine(List<Point> SameDecks)
+        {
+            bool lineFlagX = true;
+            bool lineFlagY = true;
+
+            for (int i = 0; i < SameDecks.Count - 1; i++)
+            {
+                lineFlagX &= SameDecks[i].X == SameDecks[i + 1].X;
+                lineFlagY &= SameDecks[i].Y == SameDecks[i + 1].Y;
+            }
+            return (lineFlagX ^ lineFlagY);
+        }
+        public bool ShipCoordsUnion(List<Point> SameDecks)
+        {
+            bool unionFlagX = true;
+            bool unionFlagY = true;
+            for (int i = 0; i < SameDecks.Count - 1; i++)
+            {
+                unionFlagX &= (SameDecks[i].X == SameDecks[i + 1].X) || (SameDecks[i].X == SameDecks[i + 1].X + 1) || (SameDecks[i].X == SameDecks[i + 1].X - 1);
+                unionFlagY &= (SameDecks[i].Y == SameDecks[i + 1].Y) || (SameDecks[i].Y == SameDecks[i + 1].Y + 1) || (SameDecks[i].Y == SameDecks[i + 1].Y - 1);
+            }
+            return (unionFlagX && unionFlagY);
+        }
+        public bool ShipCoordsUnique(List<Point> SameDecks)
```

Comment:
You probably could replace this method by sorting your list before equations, then ShipCoordsLine() method will validate also uniqueness. It will be much more readable, take less space and effective.

Reply:
As I understand I need to sort X coord then Y coord, but if we do this, we replacing 2 loops (each pair by all other pairs) by 2 other loops(x-sorted pairs and y-sorted) pairs. Does it really make a sense?

Reply:
You can do sorting as simply as SameDecks.OrderBy(d => d.X).ThenBy(d => d.Y).ToList() which is definitely more readable, what about complexity, yes, it will still have two loops, but as I remember it utilizes a quick sort algorithm so it still will be better than what you did. Also, it can solve for you the problem with unaligned but still valid coords like [[1,2], [1,4], [1,3]]. So I would prefer to do it that way.